### ISSUE TITLE
Document revision history

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#2956]
Sheets are IfcDocumentInformation, which seems to be semantically appropriate. That can store the latest revision, date, editors, etc. However what is not possible is to store a revision history (which usually consists of a revision ID, date, description, one or more actors, possibly a status, etc).

@aothms any ideas on what is the best way to store this?

**Comment by brunopostle** [comment]
I'm hoping that at some point this revision list can be generated automatically from tags in the Git repository, ie. this is metadata external to the IFC file itself and is never out-of-sync.

Also current revision, date and revision clouds can be automated similarly.

**Comment by Moult** [comment]
@brunopostle that would be great! Like if we knew where the sheets were published, we could use gitpython to check the gitlog  for those. There might need to be a little convention for including extra metadata apart from time/description/author.

I feel that either there should be a way to make Git seem seamless, or a non-Git fallback for non technical folks.

**Comment by brunopostle** [comment]
I wasn't thinking of versioning the generated sheets, the version tagging is at the IFC file level. Imagine having a CI server that builds the full set of documentation every time you push a commit and/or tag, and that puts these SVGs somewhere browsable.

non-git workflows need to be supported, but a git-aware sheet template can have hooks that pull this stuff directly from the repository metadata.

**Comment by Moult** [comment]
@brunopostle that would be cool, it might be possible so that every tag the system is smart enough to know which sheets have changed and can either uprev all sheets, or even know which sheets have changed and what has changed within those sheets (e.g. show relevant commits). The "revision table" on the titleblock would be a rather outdated concept in comparison.

If there is truly a more advanced future (and it sounds like there is), it would be a strong argument in favour of not storing this in IFC and instead having some sort of low-tech stopgap.

**Comment by aothms** [comment]
Exactly, there could be again a moment of reflection on whether IFC is the main umbrella or whether (a) there is some sort of folder structure with heterogeneous content that represents the project or (b) something like ICDD is used.

> the system is smart enough to know which sheets have changed and can either uprev all sheets

Maybe this paper [0] can somehow be relevant to semantically model the relationship between the drawings and the elements represented on them, so that indeed the system can determine which sheets are updated when an element changed. On the other hand, why not do it the other way around, if anyway it is a CI/CD system, why not regenerate everything and only the sheets that changed get a version bump.

[0] https://pure.tudelft.nl/ws/portalfiles/portal/104467111/w78_2021_paper_082_1.pdf

In the end monolithic IFC stuff is not scalable and in case of drawings specifically we can be pretty sure that there is not going to be any interest within bSI to advance on paper space semantics. It would be more to the open source mindset to standardize around SVG and the tooling that exists and maybe use ICDD to store the relationship between the two. It would be much more flexible.

That said, the monolithic IFC has gotten us pretty far and I also don't see any immediate pressing issues. If sheet info continues to be stored within IFC I think I'd prefer something like an IfcObject descendant as these huge non-rooted but semantic constructs are really also a relic from the past and everything under IfcObject is much more flexible and automatically has (some sort of) revision info.

It would be really cool to sketch this CI/CD system (which should probably have the possibility to be ran locally as git commit hooks). If we have this, we have a clearer picture of which is explicit data, which is derived and how to derive (git or python).
